### PR TITLE
[TK-05788 TK-05804 TK-05806 TK-05807] Kitsune Peer Discover Multi

### PR DESCRIPTION
**Pull request by neonphog**

### CODE REVIEW

File: crates/holochain/src/conductor/conductor.rs [hunk 1]
```diff
@@ -648,11 +647,34 @@ where
 
         let p2p_kv = AgentKv::new(environ.clone().into())?;
         let env = environ.guard();
-        let reader = env.reader()?;
 
-        Ok(p2p_kv
-            .as_store_ref()
-            .get(&reader, &(&*kitsune_space, &*kitsune_agent).into())?)
+        let res = {
+            let reader = env.reader()?;
```
Comment (freesig) [inline crates/holochain/src/conductor/conductor.rs:652]: One thing to keep in mind is that if new agent info is added between this read and the below `.with_commit()` that deletes the info then the new info will be lost. Basically the read will be stale.
You probably want to get the agent info and delete it in the same transaction.

File: crates/holochain/src/conductor/conductor.rs [hunk 2]
```diff
@@ -663,10 +685,44 @@ where
...
+                        }
+                    }
+                    Ok(None) => break,
+                    Err(e) => return Err(e.into()),
+                }
+            }
+        }
+
+        if !expired.is_empty() {
+            env.with_commit(|writer| {
```
Comment (freesig) [inline crates/holochain/src/conductor/conductor.rs:718]: I think this has the same issue as above.

File: crates/holochain_p2p/src/test.rs [hunk 3]
```diff
@@ -392,7 +392,7 @@ mod tests {
 
         p2p.join(dna.clone(), a1.clone()).await.unwrap();
         p2p.join(dna.clone(), a2.clone()).await.unwrap();
-        p2p.join(dna.clone(), a3.clone()).await.unwrap();
+        //p2p.join(dna.clone(), a3.clone()).await.unwrap();
```
Comment (freesig) [inline crates/holochain_p2p/src/test.rs:395]: Is this broken?

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs [hunk 4]
```diff
@@ -713,73 +642,74 @@ impl Space {
...
+                    agent: a,
+                    response: c.data.into(),
+                }),
+                _ => Err(()),
+            },
+        );
 
-            let mut out = Vec::new();
+        Ok(async move {
+            let mut out: Vec<actor::RpcMultiResponse> = futures::future::join_all(local_all)
```
Comment (freesig) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:697]: Probably not an issue now but just noting this for the future. `join_all()` will cancel all futures if any fail. Not sure if that's desired behaviour but probably fine for now.

**Review by freesig — CHANGES_REQUESTED** [review]
Looking good. 
Just think might need to change the reader to be the same transaction as the writer that deletes in those two spots.

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs (hunk 4 above)
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space.rs:697]: Fail in terms of any futures that panic!, though... which, if our code is panicking, we have bigger problems.

File: crates/holochain_p2p/src/test.rs (hunk 3 above)
Comment (neonphog) [reply crates/holochain_p2p/src/test.rs:395]:
```suggestion
```

nah, it was just easier to modify the test so it didn't include this, due to the default node count values - I'll delete that commented line.

File: crates/holochain/src/conductor/conductor.rs (hunk 1 above)
Comment (neonphog) [reply crates/holochain/src/conductor/conductor.rs]: good call, but unfortunate... now all peer requests require a write lock : ( - but we can proceed with this for now.

File: crates/holochain/src/conductor/conductor.rs (hunk 2 above)
Comment (neonphog) [reply crates/holochain/src/conductor/conductor.rs]: ditto

**Review by freesig — APPROVED** [review]

**Pull request merged**